File: sdk/typescript/tests/pangram.test.ts
/**
 * Pangram module tests — TypeScript SDK.
 *
 * Verifies cross-SDK byte-identical SHA-256 with Python and Go via the
 * EXPECTED_PANGRAM_SHA256 constant shared across all three SDKs.
 */
import { describe, it, expect } from "vitest";
import { createHash } from "crypto";

import {
  EXPECTED_PANGRAM_SHA256,
  PANGRAM_ASD_VERSION,
  PANGRAM_BODY,
  PANGRAM_MACRO_ID,
  PANGRAM_SHA256,
  PANGRAM_SHA256_TRUNCATED_16,
  PANGRAM_UTF8_BYTES,
  emit,
  emitBytes,
  macroInvocation,
  metadata,
  verifyReceived,
} from "../src/pangram.js";

describe("Pangram canonical constants", () => {
  it("body is 185 UTF-8 bytes", () => {
    expect(PANGRAM_UTF8_BYTES).toBe(185);
  });

  it("computed SHA-256 matches registered fingerprint", () => {
    expect(PANGRAM_SHA256).toBe(EXPECTED_PANGRAM_SHA256);
    expect(PANGRAM_SHA256).toBe(
      "fcefe9363ab737be174829dd8c12f4ca365fefb3601464514dd2aa4e1e0332ba",
    );
  });

  it("truncated SHA-256 is first 16 hex chars", () => {
    expect(PANGRAM_SHA256_TRUNCATED_16).toBe(PANGRAM_SHA256.slice(0, 16));
    expect(PANGRAM_SHA256_TRUNCATED_16).toBe("fcefe9363ab737be");
  });

  it("macro ID is PANGRAM", () => {
    expect(PANGRAM_MACRO_ID).toBe("PANGRAM");
  });

  it("ASD version is pinned to v15.1", () => {
    expect(PANGRAM_ASD_VERSION).toBe("v15.1");
  });
});

describe("Pangram emit functions", () => {
  it("emit returns canonical body", () => {
    expect(emit()).toBe(PANGRAM_BODY);
  });

  it("emitBytes returns UTF-8 byte buffer", () => {
    const buf = emitBytes();
    expect(buf).toEqual(Buffer.from(PANGRAM_BODY, "utf-8"));
    expect(buf.length).toBe(185);
  });
});

describe("Pangram verifyReceived", () => {
  it("accepts canonical body as string", () => {
    expect(verifyReceived(PANGRAM_BODY)).toBe(true);
  });

  it("accepts canonical body as bytes", () => {
    expect(verifyReceived(Buffer.from(PANGRAM_BODY, "utf-8"))).toBe(true);
  });

  it("rejects modified body", () => {
    expect(verifyReceived(PANGRAM_BODY + "X")).toBe(false);
  });

  it("rejects single-bit-flip in first byte", () => {
    const buf = Buffer.from(PANGRAM_BODY, "utf-8");
    buf[0] = buf[0] ^ 1;
    expect(verifyReceived(buf)).toBe(false);
  });

  it("truncated comparison accepts canonical", () => {
    expect(verifyReceived(PANGRAM_BODY, { truncated: true })).toBe(true);
  });

  it("truncated comparison rejects modified", () => {
    expect(verifyReceived(PANGRAM_BODY + "X", { truncated: true })).toBe(false);
  });
});

describe("Pangram metadata + macro invocation", () => {
  it("macroInvocation returns A:MACRO[PANGRAM]", () => {
    expect(macroInvocation()).toBe("A:MACRO[PANGRAM]");
  });

  it("metadata covers all 9 namespaces", () => {
    const m = metadata();
    expect(m.namespacesCovered).toEqual(["A", "D", "G", "H", "I", "L", "N", "R", "T"]);
    expect(m.byteLengthUtf8).toBe(185);
    expect(m.sha256).toBe(EXPECTED_PANGRAM_SHA256);
  });
});

describe("Pangram cross-SDK parity", () => {
  it("SHA-256 is byte-identical to Python and Go EXPECTED_PANGRAM_SHA256 constant", () => {
    // The Python and Go SDKs hardcode the same EXPECTED_PANGRAM_SHA256 constant
    // and verify it at module load. Re-compute here to confirm independently.
    const h = createHash("sha256")
      .update(Buffer.from(PANGRAM_BODY, "utf-8"))
      .digest("hex");
    expect(h).toBe(EXPECTED_PANGRAM_SHA256);
  });
});

// ─────────────────────────────────────────────────────────────────────────
// Tier 2 — short-form pangram (LoRa floor)
// ─────────────────────────────────────────────────────────────────────────

import {
  EXPECTED_PANGRAM_TINY_SHA256,
  PANGRAM_TINY_BODY,
  PANGRAM_TINY_MACRO_ID,
  PANGRAM_TINY_SHA256,
  PANGRAM_TINY_UTF8_BYTES,
  ChannelTier,
  emitForTier,
  verifyForTier,
} from "../src/pangram.js";

describe("Pangram tier 2 (PANGRAM_TINY)", () => {
  it("fits LoRa floor (≤ 51 UTF-8 bytes)", () => {
    expect(PANGRAM_TINY_UTF8_BYTES).toBeLessThanOrEqual(51);
  });

  it("canonical SHA-256 matches expected", () => {
    expect(PANGRAM_TINY_SHA256).toBe(EXPECTED_PANGRAM_TINY_SHA256);
    expect(PANGRAM_TINY_SHA256).toBe(
      "91c807dbbf3693ca57fb9b10ca39a5092d69de63df19b019217460e5e9c04564",
    );
  });

  it("macro id is PANGRAM_TINY", () => {
    expect(PANGRAM_TINY_MACRO_ID).toBe("PANGRAM_TINY");
  });

  it("exercises cluster claim 4 minimum primitives", () => {
    expect(PANGRAM_TINY_BODY).toContain(":"); // frame structure
    expect(PANGRAM_TINY_BODY).toContain("@"); // target syntax
    expect(PANGRAM_TINY_BODY).toContain(">"); // threshold operator
    expect(PANGRAM_TINY_BODY).toContain(";"); // sequence operator
    expect(PANGRAM_TINY_BODY).toContain("∧"); // conjunction (∧)
    expect(/[⚠↺⊘]/.test(PANGRAM_TINY_BODY)).toBe(true); // some CC designator
  });
});

describe("Pangram tier-aware API", () => {
  it("emitForTier(STANDARD) returns full pangram body", () => {
    expect(emitForTier(ChannelTier.Standard)).toBe(PANGRAM_BODY);
  });

  it("emitForTier(LoraFloor) returns short-form pangram body", () => {
    expect(emitForTier(ChannelTier.LoraFloor)).toBe(PANGRAM_TINY_BODY);
  });

  it("verifyForTier accepts canonical per-tier", () => {
    expect(verifyForTier(PANGRAM_BODY, ChannelTier.Standard)).toBe(true);
    expect(verifyForTier(PANGRAM_TINY_BODY, ChannelTier.LoraFloor)).toBe(true);
  });

  it("verifyForTier rejects cross-tier mismatch", () => {
    expect(verifyForTier(PANGRAM_BODY, ChannelTier.LoraFloor)).toBe(false);
    expect(verifyForTier(PANGRAM_TINY_BODY, ChannelTier.Standard)).toBe(false);
  });

  it("verifyForTier truncated comparison per-tier", () => {
    expect(
      verifyForTier(PANGRAM_BODY, ChannelTier.Standard, { truncated: true }),
    ).toBe(true);
    expect(
      verifyForTier(PANGRAM_TINY_BODY, ChannelTier.LoraFloor, { truncated: true }),
    ).toBe(true);
  });
});

File: sdk/go/osmp/pangram_test.go
// Pangram module tests — Go SDK.
//
// Verifies cross-SDK byte-identical SHA-256 with Python and TypeScript via
// the ExpectedPangramSHA256 constant shared across all three SDKs.

package osmp_test

import (
	"crypto/sha256"
	"encoding/hex"
	"testing"

	"github.com/octid-io/cloudless-sky/sdk/go/osmp"
)

func TestPangramBody185Bytes(t *testing.T) {
	if osmp.PangramUTF8Bytes != 185 {
		t.Errorf("expected 185 UTF-8 bytes, got %d", osmp.PangramUTF8Bytes)
	}
}

func TestPangramSHA256MatchesExpected(t *testing.T) {
	if osmp.PangramSHA256 != osmp.ExpectedPangramSHA256 {
		t.Errorf("computed %s != expected %s", osmp.PangramSHA256, osmp.ExpectedPangramSHA256)
	}
	want := "fcefe9363ab737be174829dd8c12f4ca365fefb3601464514dd2aa4e1e0332ba"
	if osmp.PangramSHA256 != want {
		t.Errorf("PangramSHA256 = %s, want %s", osmp.PangramSHA256, want)
	}
}

func TestPangramTruncatedSHA256(t *testing.T) {
	if osmp.PangramSHA256Truncated16 != osmp.PangramSHA256[:16] {
		t.Errorf("truncated %s != first-16 of full %s", osmp.PangramSHA256Truncated16, osmp.PangramSHA256[:16])
	}
	if osmp.PangramSHA256Truncated16 != "fcefe9363ab737be" {
		t.Errorf("truncated SHA = %s, want fcefe9363ab737be", osmp.PangramSHA256Truncated16)
	}
}

func TestPangramMacroID(t *testing.T) {
	if osmp.PangramMacroID != "PANGRAM" {
		t.Errorf("macro id = %q, want PANGRAM", osmp.PangramMacroID)
	}
}

func TestPangramASDVersion(t *testing.T) {
	if osmp.PangramASDVersion != "v15.1" {
		t.Errorf("ASD version = %q, want v15.1", osmp.PangramASDVersion)
	}
}

func TestEmitPangram(t *testing.T) {
	if osmp.EmitPangram() != osmp.PangramBody {
		t.Errorf("EmitPangram returned non-canonical body")
	}
}

func TestEmitPangramBytes(t *testing.T) {
	b := osmp.EmitPangramBytes()
	if string(b) != osmp.PangramBody {
		t.Errorf("EmitPangramBytes returned non-canonical bytes")
	}
	if len(b) != 185 {
		t.Errorf("EmitPangramBytes returned %d bytes, want 185", len(b))
	}
}

func TestVerifyReceivedCanonical(t *testing.T) {
	if !osmp.VerifyReceivedString(osmp.PangramBody, false) {
		t.Errorf("canonical body did not verify")
	}
	if !osmp.VerifyReceived(osmp.EmitPangramBytes(), false) {
		t.Errorf("canonical bytes did not verify")
	}
}

func TestVerifyReceivedModified(t *testing.T) {
	if osmp.VerifyReceivedString(osmp.PangramBody+"X", false) {
		t.Errorf("modified body should not verify")
	}
	bad := []byte(osmp.PangramBody)
	bad[0] = bad[0] ^ 1
	if osmp.VerifyReceived(bad, false) {
		t.Errorf("bit-flipped first byte should not verify")
	}
}

func TestVerifyReceivedTruncated(t *testing.T) {
	if !osmp.VerifyReceivedString(osmp.PangramBody, true) {
		t.Errorf("canonical body should verify with truncated hash")
	}
	if osmp.VerifyReceivedString(osmp.PangramBody+"X", true) {
		t.Errorf("modified body should not verify with truncated hash")
	}
}

func TestPangramMacroInvocation(t *testing.T) {
	got := osmp.PangramMacroInvocation()
	want := "A:MACRO[PANGRAM]"
	if got != want {
		t.Errorf("PangramMacroInvocation = %q, want %q", got, want)
	}
}

func TestPangramMetadata(t *testing.T) {
	m := osmp.PangramMetadataInfo()
	if m.MacroID != "PANGRAM" || m.ByteLengthUTF8 != 185 || m.ASDVersion != "v15.1" {
		t.Errorf("metadata structural fields wrong: %+v", m)
	}
	wantNS := []string{"A", "D", "G", "H", "I", "L", "N", "R", "T"}
	if len(m.NamespacesCovered) != len(wantNS) {
		t.Errorf("namespaces covered = %d, want %d", len(m.NamespacesCovered), len(wantNS))
	}
	for i, ns := range wantNS {
		if m.NamespacesCovered[i] != ns {
			t.Errorf("namespace[%d] = %q, want %q", i, m.NamespacesCovered[i], ns)
		}
	}
}

func TestPangramCrossSDKParity(t *testing.T) {
	// Re-compute SHA-256 here using Go's stdlib and compare to the registered
	// constant. The TypeScript and Python SDKs do the same against the same
	// EXPECTED_PANGRAM_SHA256; if all three match, byte-identicalness holds.
	sum := sha256.Sum256([]byte(osmp.PangramBody))
	got := hex.EncodeToString(sum[:])
	if got != osmp.ExpectedPangramSHA256 {
		t.Errorf("cross-SDK parity broken: re-computed %s != expected %s", got, osmp.ExpectedPangramSHA256)
	}
}

// ─────────────────────────────────────────────────────────────────────────
// Tier 2 — short-form pangram (LoRa floor)
// ─────────────────────────────────────────────────────────────────────────

func TestPangramTinyFitsLoraFloor(t *testing.T) {
	if osmp.PangramTinyUTF8Bytes > 51 {
		t.Errorf("PANGRAM_TINY (%dB) violates LoRa floor 51B invariant", osmp.PangramTinyUTF8Bytes)
	}
}

func TestPangramTinyCanonicalSHA256(t *testing.T) {
	if osmp.PangramTinySHA256 != osmp.ExpectedPangramTinySHA256 {
		t.Errorf("computed %s != expected %s", osmp.PangramTinySHA256, osmp.ExpectedPangramTinySHA256)
	}
	want := "91c807dbbf3693ca57fb9b10ca39a5092d69de63df19b019217460e5e9c04564"
	if osmp.PangramTinySHA256 != want {
		t.Errorf("PangramTinySHA256 = %s, want %s", osmp.PangramTinySHA256, want)
	}
}

func TestPangramTinyMacroID(t *testing.T) {
	if osmp.PangramTinyMacroID != "PANGRAM_TINY" {
		t.Errorf("macro id = %q, want PANGRAM_TINY", osmp.PangramTinyMacroID)
	}
}

func TestEmitForTier(t *testing.T) {
	if osmp.EmitForTier(osmp.ChannelTierStandard) != osmp.PangramBody {
		t.Errorf("EmitForTier(Standard) returned non-canonical body")
	}
	if osmp.EmitForTier(osmp.ChannelTierLoraFloor) != osmp.PangramTinyBody {
		t.Errorf("EmitForTier(LoraFloor) returned non-canonical short-form body")
	}
}

func TestVerifyForTierCanonical(t *testing.T) {
	if !osmp.VerifyForTierString(osmp.PangramBody, osmp.ChannelTierStandard, false) {
		t.Errorf("standard body did not verify against standard tier")
	}
	if !osmp.VerifyForTierString(osmp.PangramTinyBody, osmp.ChannelTierLoraFloor, false) {
		t.Errorf("tiny body did not verify against LoRa-floor tier")
	}
}

func TestVerifyForTierCrossRejects(t *testing.T) {
	if osmp.VerifyForTierString(osmp.PangramBody, osmp.ChannelTierLoraFloor, false) {
		t.Errorf("standard body should not verify against LoRa-floor tier")
	}
	if osmp.VerifyForTierString(osmp.PangramTinyBody, osmp.ChannelTierStandard, false) {
		t.Errorf("tiny body should not verify against standard tier")
	}
}

func TestPangramTinyMinimumPrimitives(t *testing.T) {
	body := osmp.PangramTinyBody
	for _, p := range []string{":", "@", ">", ";", "∧"} {
		if !contains(body, p) {
			t.Errorf("PANGRAM_TINY missing required primitive %q", p)
		}
	}
	hasCC := false
	for _, cc := range []string{"⚠", "↺", "⊘"} {
		if contains(body, cc) {
			hasCC = true
			break
		}
	}
	if !hasCC {
		t.Errorf("PANGRAM_TINY missing consequence class designator")
	}
}

func contains(s, sub string) bool {
	for i := 0; i+len(sub) <= len(s); i++ {
		if s[i:i+len(sub)] == sub {
			return true
		}
	}
	return false
}

func TestPangramTinyCrossSDKParity(t *testing.T) {
	sum := sha256.Sum256([]byte(osmp.PangramTinyBody))
	got := hex.EncodeToString(sum[:])
	if got != osmp.ExpectedPangramTinySHA256 {
		t.Errorf("tier 2 cross-SDK parity broken: %s != %s", got, osmp.ExpectedPangramTinySHA256)
	}
}
